<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Project 4 - Unveiling Chronic Disease in Singaporean Lifestyle

> Authors: Chung Yau, Gilbert, Han Kiong, Zheng Gang
---

**Problem Statement:**  
In Singapore, the increasing prevalence of chronic diseases presents a pressing public health concern, underscoring the need for proactive intervention strategies. 

How can we identify individuals at high risk for chronic diseases based on their behavioral habits? By doing so, we can enable early detection and provide recommendations, fostering a proactive approach to preventing various chronic diseases.

  
**Target Audience:**  
Product team in Synapxe, in preparation for Healthier SG 2025 roadmap workshop. 

These are the notebooks for this project:  
 1. `01_Data_Collection_Food.ipynb`  
 2. `02_Data_Preprocessing.ipynb`   
 3. `03_FeatureEngineering_and_EDA.ipynb`
 4. `04_Data_Modelling.ipynb` 
 5. `05_Hyperparameter_Model Fitting_Evaluation.ipynb`
 6. `05a_Model_Pickling.ipynb`
 7. `06_Implementation_FoodRecommender.ipynb` 

 ---

# This Notebook: 02_Data_Preprocessing
- We will focus on cleaning the dataset and dropping irrelevant features.
- The expected output of this notebook will be a file that is suitable for upcoming EDA (Exploratory Data Analysis) and modeling in subsequent notebooks.

---

### 1. Import libraries

Imports necessary libraries for data manipulation in Python.

In [3]:
import pandas as pd

---
### 2. Read CSV

In [4]:
# pandas to read the data file csv which was zipped.
dataframe = pd.read_csv("../data/2015.zip", compression='zip')

In [5]:
dataframe.head(5)

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE         SEQNO  \
0     1.0     1.0  b'01292015'  b'01'  b'29'  b'2015'    1200.0  2.015000e+09   
1     1.0     1.0  b'01202015'  b'01'  b'20'  b'2015'    1100.0  2.015000e+09   
2     1.0     1.0  b'02012015'  b'02'  b'01'  b'2015'    1200.0  2.015000e+09   
3     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   
4     1.0     1.0  b'01142015'  b'01'  b'14'  b'2015'    1100.0  2.015000e+09   

           _PSU  CTELENUM  ...  _PAREC1  _PASTAE1  _LMTACT1  _LMTWRK1  \
0  2.015000e+09       1.0  ...      4.0       2.0       1.0       1.0   
1  2.015000e+09       1.0  ...      2.0       2.0       3.0       3.0   
2  2.015000e+09       1.0  ...      9.0       9.0       9.0       9.0   
3  2.015000e+09       1.0  ...      4.0       2.0       1.0       1.0   
4  2.015000e+09       1.0  ...      4.0       2.0       1.0       1.0   

   _LMTSCL1  _RFSEAT2  _RFSEAT3  _FLSHOT6  _PNEUMO2  _AIDTST3  
0       1.0       1.0       1.0       NaN       NaN       1.0  
1       4.0       2.0       2.0       NaN       NaN       2.0  
2       9.0       9.0       9.0       9.0       9.0       NaN  
3       1.0       1.0       1.0       NaN       NaN       9.0  
4       1.0       1.0       1.0       NaN       NaN       1.0  

[5 rows x 330 columns]

Checking the shape of our dataset

In [6]:
dataframe.shape

(441456, 330)

There are 330 columns in the dataset. We will drop irrelevant features and focus on features that we wish to investigate in the next section. 

---
### 3. Dropping irrelevant features

We will be focusing on the following categories of features for our analysis:   
1. Alcohol consumption habits, 
2. Existing health conditions, 
3. User demographics,  
4. Physical activity habits,   
5. Smoking habits, and   
6. Dietary habits.

Due to the sheer volume of columns. The team finds the need to manually sieve through feature by feature to determine the suitability of the columns. For specifics of how we determine which column to drop can be found in this gsheet. The columns that are marked `"To include == Y"` are those that are identified as the candidate features. 
https://docs.google.com/spreadsheets/d/1jVtQ1TYKtLA9SFqdjPmhYl1fqRrUywjBrI5U5cKXbBo/edit#gid=1776346838. A copy of the analysis file (`Variable list.xlsx`) is available in the data/ folder.

In the data preprocessing, the features that have high percentage (i.e. 5%) of empty values, as well as column from the column pair with strong correlation ( correlation coefficient > 0.8) will be dropped. The filtered features will then be cross checked against the candidate features from the analysis. Discrepancies, if any, will be dealt with accordingly. 

BLANK in this dataset implies that the data is either missing or not asked.

 

We will first proceed to drop features with more than 5% of the respondants (20,000 missing values) for our first cut. 

In [7]:
null_counts = dataframe.isnull().sum()  
sorted_null_counts = null_counts.sort_values(ascending=False) 
sorted_null_counts_df = pd.DataFrame({
    "feature": sorted_null_counts.index,
    "null_count": sorted_null_counts.values  
})

filtered_sorted_null_counts_df = sorted_null_counts_df[sorted_null_counts_df["null_count"] >= 20000]
filtered_sorted_null_counts_df["feature"].values
columns_to_drop_v1 = filtered_sorted_null_counts_df["feature"].values

In [8]:
dataframe_v1 = dataframe.drop(columns = columns_to_drop_v1)

In [9]:
dataframe_v1.shape

(441456, 121)

We now have 121 columns. we now manually assess the features and drop those that are not of our interest above or we already see that there are redundant features that are already represented by other columns.

In [10]:
columns_to_drop_v2 = [
    'USENOW3', 'INTERNET', 'INCOME2', '_LMTACT1', '_LMTWRK1', '_LMTSCL1', '_STRWT', 'QSTLANG', 'VETERAN3', 'CHILDREN',
    'MEDCOST', 'CHECKUP1', 'HLTHPLN1', '_STATE', 'FMONTH', 'IDATE', 'IMONTH', 'IDAY', 'IYEAR', 'DISPCODE', 'SEQNO',
    '_PSU', 'PERSDOC2', 'RENTHOM1', 'PCDMDECN', 'QSTVER', '_STSTR', '_RAWRAKE', '_WT2RAKE', '_DUALUSE', '_HCVU651',
    '_CHOLCHK', '_CHLDCNT', '_MISFRTN', '_MISVEGN', '_FRTRESP', '_VEGRESP', '_FRUITEX', '_VEGETEX', 'MAXVO2_', 'FC60_',
    'PAMISS1_', '_PACAT1', '_PASTRNG', '_PAREC1', '_PASTAE1', '_RFSEAT2', '_RFSEAT3', 'EXACTOT1', 'EXACTOT2', '_INCOMG', '_PA300R2'
]

In [11]:
dataframe_v2 = dataframe_v1.drop(columns = columns_to_drop_v2)

In [12]:
dataframe_v2.shape

(441456, 69)

We are left with 69 features from the original 330. we now run a correlation analysis to identify columns with correlation more than 0.8. This indicates repeated information and we can further simpilfy our dataset with this. 

In [13]:
correlation_matrix = dataframe_v2.corr()

high_correlation = correlation_matrix[correlation_matrix > 0.5]
high_correlation = high_correlation.unstack().reset_index()
high_correlation.columns = ['Column 1', 'Column 2', 'Correlation']

high_correlation = high_correlation[(high_correlation['Column 1'] != high_correlation['Column 2']) & 
                                    (high_correlation['Correlation'].notna())]

high_correlation = high_correlation.drop_duplicates(subset=['Correlation'])

high_correlation = high_correlation[high_correlation['Correlation'] > 0.80]

print("Pairs of columns with correlation greater than 0.8")
print(high_correlation)

Pairs of columns with correlation greater than 0.8
      Column 1  Column 2  Correlation
866   HAVARTH3  _DRDXAR1     1.000000
1295     EDUCA   _EDUCAG     0.979974
1432   WEIGHT2   _RFBMI5     0.860003
2055  SMOKE100  _SMOKER3     0.802848
2731   _PRACE1   _MRACE1     0.998638
2942     _RACE  _RACEGR3     0.957102
2943     _RACE  _RACE_G1     0.835171
3011  _RACEG21  _RACEGR3     0.835251
3012  _RACEG21  _RACE_G1     0.883616
3081  _RACEGR3  _RACE_G1     0.858784
3222  _AGEG5YR    _AGE80     0.976520
3223  _AGEG5YR    _AGE_G     0.951450
3361    _AGE80    _AGE_G     0.971136
3501     HTIN4      HTM4     0.999611
3921  DRNKANY5  DROCDY3_     0.916879
3922  DRNKANY5  _RFBING5     0.826619
3923  DRNKANY5  _DRNKWEK     0.851606
3924  DRNKANY5  _RFDRHV5     0.841366
3991  DROCDY3_  _RFBING5     0.907486
3992  DROCDY3_  _DRNKWEK     0.909973
3993  DROCDY3_  _RFDRHV5     0.909930
4061  _RFBING5  _DRNKWEK     0.885880
4062  _RFBING5  _RFDRHV5     0.888252
4131  _DRNKWEK  _RFDRHV5     0.995035

Based on the above, we then drop more features for those which are severely correlated. 

In [14]:
columns_to_drop_v3 = [
    'DIFFALON', 'WEIGHT2', 'BPHIGH4', 'CVDCRHD4', 'HAVARTH3', '_LLCPWT', '_RACEGR3', '_FRT16', '_VEG23', 'BLOODCHO',
    '_RFSMOK3', 'HTIN4', 'SMOKE100', 'DIFFDRES', 'DECIDE', 'USEEQUIP', 'QLACTLM2', '_RACE_G1', 'HEIGHT3',
    'GENHLTH', 'PHYSHLTH', 'EDUCA', '_MRACE1', '_HISPANC', '_RACEG21', '_AGE65YR', '_AGE80', '_AGE_G', '_RFBMI5',
    '_PA30021', '_TOTINDA', 'MENTHLTH', '_LTASTH1', '_CASTHM1', '_ASTHMS1', '_RFHLTH', '_PRACE1', '_PAINDX1', 'DROCDY3_', '_RFDRHV5'
]

In [15]:
dataframe_v3 = dataframe_v2.drop(columns = columns_to_drop_v3)

In [16]:
dataframe_v3.shape

(441456, 29)

In [17]:
dataframe_v3.columns

Index(['CVDINFR4', 'CVDSTRK3', 'ASTHMA3', 'CHCSCNCR', 'CHCOCNCR', 'CHCCOPD1',
       'ADDEPEV2', 'CHCKIDNY', 'DIABETE3', 'SEX', 'MARITAL', 'EMPLOY1',
       'BLIND', 'DIFFWALK', 'ALCDAY5', '_RFHYPE5', '_MICHD', '_DRDXAR1',
       '_RACE', '_AGEG5YR', 'HTM4', '_EDUCAG', '_SMOKER3', 'DRNKANY5',
       '_RFBING5', '_DRNKWEK', '_FRTLT1', '_VEGLT1', '_PA150R2'],
      dtype='object')

Proceed to verify whether any discrepancy with the candidate features identified during analysis

> The `Variable list - Tentative list.csv` is the analysis of the data based on given data dictionary from the source. A group of columns that are marked "To include = Y" are the candidate columns to be used for Features based on our analysis. 

In [18]:
feature_candidate = pd.read_csv("../data/Variable list - Tentative list.csv", skiprows=2)

In [19]:
feature_candidate.tail(5)

Feature                                        Description To include
26  _SMOKER3  Four-level smoker status: Everyday smoker, Som...          Y
27  DRNKANY5  Adults who reported having had at least one dr...          Y
28  _RFBING5  Binge drinkers (males having five or more drin...          Y
29     _RACE                          Race/ethnicity categories          Y
30   ALCDAY5  During the past 30 days, how many days per wee...          Y

In [20]:
candidate_list = feature_candidate[feature_candidate['To include'] == 'Y']['Feature']

For each of the candidate feature, check if the candidate feature is available too in the filtered feature list. If not available, captured in the `candidate_drop`, will be further dealt with. 

In [21]:
candidate_drop = []
for candidate in list(candidate_list):
    if candidate not in list(dataframe_v3):
        candidate_drop.append(candidate)

candidate_drop

['_RFCHOL', 'WTKG3']

We observed that we had intendeded to include  `_RFCHOL` and `WTKG3` in our initial analysis but they were drop. They were dropped due to high number of missing values in the datasets. The team look a deeper look and have decided the on the following:  
  
`WTKG3` - this defines the weight of the individual and approx 7.5% of the values were missing. The team sees this as an important variable to determine the risk of chronic disease for an individual and have decided to add it into the dataset. The missing values will be imputed with the missing values.   
 `_RFCHOL` - this defines if the individual has high cholestrol and approx 12.5% of the values were missing. Similarly, this is important and we will include this information while taking missing values as non-declarance of the individual having said condition.


Add both variables back to the dataset. 

In [22]:
dataframe_v4 = pd.concat([dataframe_v3, dataframe["_RFCHOL"]], axis = 1)
dataframe_v4 = pd.concat([dataframe_v4, dataframe["WTKG3"]], axis = 1)

All the candidate features identified during analysis of the raw data are available in the filtered features. Analysis outcome was aligned with the feature filtering implemented. 

Based on the above, we arrived at the 31 columns for our data cleaning and next steps.  

#### 3b. Merging 2013.csv

After the EDA, it was finalized that, out of the 441456 rows of data in the 2015.csv, only 9759 rows of data for "Asian" people are relevant for modeling. In order to increase the number of data rows for EDA and modeling, the data in the 2013.csv will be loaded and concatenate with the 2015.csv data. 

***Reason for choosing 2013.csv***
Analysis was performed on the downloaded dataset. There are total of 5 CSV provided by the data source, where every CSV contains the data collected for a particular year. The 5 CSV file, namely 2011.csv, 2012.csv, 2013.csv, 2014.csv and 2015.csv, are therefore the data collected in year 2011, 2012, 2013, 2014 and 2015 respectively. Not all the data columns are identical in the 5 files.   
  
The analysis report `Variable list - data availability.csv` listed the columns discrepencies among the 31 data columns identified for EDA and modeling, with data columns in 2015.csv as the baseline. From the analysis, 2013.csv has the least column discrepancy, with only 2 columns missing. The team decided, therefore, to concatenate the data in 2013.csv, so that additional 9510 records (for Asian people) are added to the dataframe. 

In [23]:
df_2013 = pd.read_csv("../data/2013.zip", compression='zip')
df_2013.head(5)

_STATE  FMONTH        IDATE IMONTH   IDAY    IYEAR  DISPCODE         SEQNO  \
0     1.0     1.0  b'01092013'  b'01'  b'09'  b'2013'    1100.0  2.013001e+09   
1     1.0     1.0  b'01192013'  b'01'  b'19'  b'2013'    1100.0  2.013001e+09   
2     1.0     1.0  b'01192013'  b'01'  b'19'  b'2013'    1100.0  2.013001e+09   
3     1.0     1.0  b'01112013'  b'01'  b'11'  b'2013'    1100.0  2.013001e+09   
4     1.0     2.0  b'02062013'  b'02'  b'06'  b'2013'    1100.0  2.013001e+09   

           _PSU  CTELENUM  ...  _AGE80  _IMPEDUC  _IMPMRTL  _IMPHOME  \
0  2.013001e+09       1.0  ...    60.0       6.0       2.0       1.0   
1  2.013001e+09       1.0  ...    50.0       5.0       1.0       1.0   
2  2.013001e+09       1.0  ...    55.0       6.0       1.0       1.0   
3  2.013001e+09       1.0  ...    64.0       4.0       1.0       1.0   
4  2.013001e+09       1.0  ...    66.0       6.0       1.0       1.0   

   RCSBRAC1  RCSRACE1  RCHISLA1  RCSBIRTH  TYPEINDS  TYPEWORK  
0       NaN       b''       b''       b''       b''       b''  
1       NaN       b''       b''       b''       b''       b''  
2       NaN       b''       b''       b''       b''       b''  
3       NaN       b''       b''       b''       b''       b''  
4       NaN       b''       b''       b''       b''       b''  

[5 rows x 336 columns]

In [24]:
df_2013.shape

(491773, 336)

Check the column discrepancy between 2013 dataset and 2015 dataset. Note that 2015 dateset is the baseline dataset where the 31 columns were identified from. 

In [25]:
columns_2015df = set(dataframe_v4.columns)
columns_2013df = set(df_2013.columns) 
columns_not_in_2013df = columns_2015df - columns_2013df
columns_not_in_2013df

{'_DRNKWEK', '_MICHD'}

We see that `_DRNKWEK` and `_MICHD` is not in the 2013 dataset. However, there are similar columns within the same year that we can use.  

`_DRNKWEK` : represents the total number of alcoholic beverages consumed per week. It is calculated by taking `AVEDRNK2` * `DROCDY3_` *7. Both `AVEDRNK2`, `DROCDY3_` are present in the 2013 dataset. Thus, we will be able to calculate them. 

`_MICHD` : represents if the users have coronary heart diesease (CHD) or myocardial infarction (MI). It is possible to derive these information from `CVDINFR4` and `CVDCRHD4`. We will be able to create these columns. 

In [26]:
df_2013["DROCDY3_"] = df_2013["DROCDY3_"].map(lambda x : x if x in range(1,900) else 0)
df_2013["AVEDRNK2"] = df_2013["AVEDRNK2"].map(lambda x : x if x in range(1,77) else 0)
df_2013["_DRNKWEK"] = df_2013["AVEDRNK2"]*df_2013["DROCDY3_"]*7

In [27]:
df_2013["CVDINFR4"] = df_2013["CVDINFR4"].map(lambda x : 1 if x == 1 else 0)
df_2013["CVDCRHD4"] = df_2013["CVDCRHD4"].map(lambda x : 1 if x == 1 else 0)
df_2013["_MICHD"] = df_2013.apply(lambda row: max(row["CVDINFR4"], row["CVDCRHD4"]), axis=1)

Based on the 31 columns identified in 2015 dataset (in dataframe_v4) to filter out the same columns in 2013 dataset. 

In [28]:
df_2013_match = df_2013[dataframe_v4.columns].copy()

In [29]:
df_2013_match.shape

(491773, 31)

In [30]:
dataframe_v4.shape

(441456, 31)

Final confirm that both 2013 dataset and 2015 dataset has identical columns. 

In [31]:
columns_2015df = set(dataframe_v4.columns)
columns_2013df = set(df_2013_match.columns) 
columns_not_in_2013df = columns_2015df - columns_2013df
columns_not_in_2013df

set()

In [32]:
columns_not_in_2015df = columns_2013df - columns_2015df
columns_not_in_2015df

set()

The dataframe from 2013 dataset is now confirm matching the dataset from 2015. Ready to concatenate both dataset.

note: The na values in the respective columns will be dealt with in the next section, altogether after concatenate the two dataset

In [33]:
dataframe_2013_2015 = pd.concat([dataframe_v4, df_2013_match])
dataframe_2013_2015.reset_index(drop=True, inplace=True)
dataframe_2013_2015.shape

(933229, 31)

Reassign the dataframe to dataframe_v4 for continuation of the code in next section on Data Cleaning


In [34]:
dataframe_v4 = dataframe_2013_2015.copy()

#### 4. Data Cleaning

We will proceed to clean our data to be suitable for analysis in this section. We will perform actions on almost all of our columns based the google sheet here: 
https://docs.google.com/spreadsheets/d/1jVtQ1TYKtLA9SFqdjPmhYl1fqRrUywjBrI5U5cKXbBo/edit#gid=51570131 . A copy of the analysis file (`Variable list.xlsx`) is available in the data/ folder.

For easy readability, we will do the actions on a feature level with explanation. 

In [35]:
dataframe_v4.head()

CVDINFR4  CVDSTRK3  ASTHMA3  CHCSCNCR  CHCOCNCR  CHCCOPD1  ADDEPEV2  \
0       2.0       2.0      1.0       2.0       2.0       1.0       1.0   
1       2.0       2.0      2.0       2.0       2.0       2.0       2.0   
2       7.0       1.0      2.0       2.0       1.0       2.0       2.0   
3       2.0       2.0      2.0       2.0       1.0       2.0       1.0   
4       2.0       2.0      2.0       2.0       2.0       2.0       2.0   

   CHCKIDNY  DIABETE3  SEX  ...  _EDUCAG  _SMOKER3  DRNKANY5  _RFBING5  \
0       2.0       3.0  2.0  ...      2.0       3.0       2.0       1.0   
1       2.0       3.0  2.0  ...      4.0       1.0       2.0       1.0   
2       2.0       3.0  2.0  ...      2.0       9.0       9.0       9.0   
3       2.0       3.0  2.0  ...      2.0       4.0       2.0       1.0   
4       2.0       3.0  2.0  ...      3.0       4.0       2.0       1.0   

       _DRNKWEK  _FRTLT1  _VEGLT1  _PA150R2  _RFCHOL    WTKG3  
0  5.397605e-79      2.0      1.0       3.0      2.0  12701.0  
1  5.397605e-79      2.0      2.0       1.0      1.0   7484.0  
2  9.990000e+04      9.0      9.0       9.0      2.0   7167.0  
3  5.397605e-79      1.0      2.0       3.0      2.0   8165.0  
4  5.397605e-79      9.0      1.0       3.0      1.0   6441.0  

[5 rows x 31 columns]

In [36]:

# Column name: _RFCHOL
# Replace missing values with 0. Standardardize response for analysis.

dataframe_v4["_RFCHOL"] = dataframe_v4["_RFCHOL"].fillna(1)
dataframe_v4["_RFCHOL"] = dataframe_v4["_RFCHOL"].map( lambda x : 1 if x == 2 else 0)

# Column name: WTKG3
# Replace missing values with mean weight
mean_weight = dataframe_v4["WTKG3"].mean()
dataframe_v4["WTKG3"] = dataframe_v4["WTKG3"].fillna(mean_weight)

# Column name: HTM4
# Replace missing values with mean height
mean_height = dataframe_v4["HTM4"].mean()
dataframe_v4["HTM4"] = dataframe_v4["HTM4"].fillna(mean_height)

# Column name: DIFFWALK
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["DIFFWALK"] = dataframe_v4["DIFFWALK"].fillna(2)
dataframe_v4["DIFFWALK"] = dataframe_v4["DIFFWALK"].map( lambda x : 1 if x == 1 else 0)

# Column name: BLIND
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["BLIND"] = dataframe_v4["BLIND"].fillna(2)
dataframe_v4["BLIND"] = dataframe_v4["BLIND"].map( lambda x : 1 if x == 1 else 0)


# Column name: _MICHD 
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["_MICHD"] = dataframe_v4["_MICHD"].fillna(2)
dataframe_v4["_MICHD"] = dataframe_v4["_MICHD"].map( lambda x : 1 if x == 1 else 0)

# Column name: _DRDXAR1
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["_DRDXAR1"] = dataframe_v4["_DRDXAR1"].fillna(2)
dataframe_v4["_DRDXAR1"] = dataframe_v4["_DRDXAR1"].map( lambda x : 1 if x == 1 else 0)

# Column name: DIABETE3
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["DIABETE3"] = dataframe_v4["DIABETE3"].fillna(3)
dataframe_v4["DIABETE3"] = dataframe_v4["DIABETE3"].map( lambda x : 1 if x < 3 else 0)

# Column name: CHCSCNCR
# Assume no for null values, define uncertain and refusal response as no 
dataframe_v4["CHCSCNCR"] = dataframe_v4["CHCSCNCR"].fillna(2)
dataframe_v4["CHCSCNCR"] = dataframe_v4["CHCSCNCR"].map( lambda x : 1 if x == 1 else 0)

# Column name: _DRNKWEK 
# Assume column mean for 99900 (no data) 
mean_drinks = dataframe_v4["_DRNKWEK"].mean()
dataframe_v4["_DRNKWEK"] = dataframe_v4["_DRNKWEK"].replace(99900, mean_drinks)

# Column name: SEX
# 1 as male. 0 as female.
dataframe_v4["SEX"] = dataframe_v4["SEX"].map( lambda x : 1 if x == 1 else 0)

# Column name: MARITAL
# Define married as married. Others classify as not married (0).
dataframe_v4["MARITAL"] = dataframe_v4["MARITAL"].map( lambda x : 1 if x == 1 else 0)

# Column name: EMPLOY1
# Simplified to only employed (1) and others
dataframe_v4["EMPLOY1"] = dataframe_v4["EMPLOY1"].map( lambda x : 1 if x in [1,2] else 0)

# Column name: _AGEG5YR
# Impute mean age category to people who were unwilling to disclose age
mean_age_group = int(dataframe_v4["_AGEG5YR"].mean())
dataframe_v4["_AGEG5YR"] = dataframe_v4["_AGEG5YR"].map( lambda x : mean_age_group if x == 14 else x)

# Column name: _EDUCAG
# Simplify education to three categories
# 0 - Did not graduate high school
# 1 - high school graduate
# 2 - College or Tech school grad
dataframe_v4["_EDUCAG"] = dataframe_v4["_EDUCAG"].map( lambda x : 0 if x in [1,9] else (1 if x in [2,3] else 2))  

# Column name: _PA150R2
# Simplify category to three categories. Impute 9 to 2.
# 0 - 0 minutes of vigorous exercise
# 1 - 1 to 149 minutes of vigorous exercise
# 2 - more than 150 minutes of vigorous exercise
dataframe_v4["_PA150R2"] = dataframe_v4["_PA150R2"].map(lambda x : 0 if x == 3 else (2 if x == 1 else 1))

# Column name: _FRTLT1
# Simplify category to two categories. 
# 0 - consumed fruits less than one time per day
# 1 - consume fruits more than one time per day
dataframe_v4["_FRTLT1"] = dataframe_v4["_FRTLT1"].map(lambda x : 1 if x == 1 else 0)

# Column name: _VEGLT1
# Simplify category to two categories.
# 0 - consumed veg less than one time per day
# 1 - consume veg more than one time per day
dataframe_v4["_VEGLT1"] = dataframe_v4["_VEGLT1"].map(lambda x : 1 if x == 1 else 0)

# Column name: CVDINFR4
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["CVDINFR4"] = dataframe_v4["CVDINFR4"].map(lambda x : 1 if x == 1 else 0)

# Column name: CVDSTRK3
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["CVDSTRK3"] = dataframe_v4["CVDSTRK3"].map(lambda x : 1 if x == 1 else 0)

# Column name: ASTHMA3
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["ASTHMA3"] = dataframe_v4["ASTHMA3"].map(lambda x : 1 if x == 1 else 0)

# Column name: CHCOCNCR
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["CHCOCNCR"] = dataframe_v4["CHCOCNCR"].map(lambda x : 1 if x == 1 else 0)

# Column name: CHCCOPD1
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["CHCCOPD1"] = dataframe_v4["CHCCOPD1"].map(lambda x : 1 if x == 1 else 0)

# Column name: ADDEPEV2
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["ADDEPEV2"] = dataframe_v4["ADDEPEV2"].map(lambda x : 1 if x == 1 else 0)

# Column name: CHCKIDNY
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["CHCKIDNY"] = dataframe_v4["CHCKIDNY"].map(lambda x : 1 if x == 1 else 0)

# Column name: _RFHYPE5
# Simplify category to two categories.
# 0 - No and others
# 1 - Yes
dataframe_v4["_RFHYPE5"] = dataframe_v4["_RFHYPE5"].map(lambda x : 1 if x == 2 else 0)

# Column name: _SMOKER3
# Map those who refuse to answer and missing data to non-smoker. 
# Revise response variable to oridinal for easy analysis. 
# 0 - never smoke
# 1 - former smoker
# 2 - current smoker (smoke some days)
# 3 - current smoker (smoke every day)
dataframe_v4["_SMOKER3"] = dataframe_v4["_SMOKER3"].map(lambda x : 3 if x == 1 else ( 2 if x == 2 else (1 if x == 3 else 0)))

# Column name: DRNKANY5
# Map those who refuse to answer as Yes and missing data as No.
# 0 - No and others
# 1 - Yes
dataframe_v4["DRNKANY5"] = dataframe_v4["DRNKANY5"].map(lambda x : 1 if x == 7 else (2 if x == 9 else x))
dataframe_v4["DRNKANY5"] = dataframe_v4["DRNKANY5"].map(lambda x : 1 if x == 1 else 0)

# Column name: _RFBING5
# Map those who refuse to answer and missing data as No.
# 0 - No and others
# 1 - Yes
dataframe_v4["_RFBING5"] = dataframe_v4["_RFBING5"].map(lambda x : 1 if x == 2 else 0)

# Column name: ALCDAY5
# we redefine feature to days per month
dataframe_v4["ALCDAY5"] = dataframe_v4["ALCDAY5"].map(lambda x : 0 if x in [777,888,999] else x)
dataframe_v4["ALCDAY5"] = dataframe_v4["ALCDAY5"].map(lambda x : (x*4)-400 if x in range(100,199) else x)
dataframe_v4["ALCDAY5"] = dataframe_v4["ALCDAY5"].map(lambda x : x-200 if x in range (200,299) else x)
mean_days_drink_month = dataframe_v4["ALCDAY5"].mean()
dataframe_v4["ALCDAY5"] = dataframe_v4["ALCDAY5"].fillna(mean_days_drink_month)

Field level transformation is done. Proceed to clean up NA data. 

In [37]:
dataframe_v4.isnull().sum()

CVDINFR4     0
CVDSTRK3     0
ASTHMA3      0
CHCSCNCR     0
CHCOCNCR     0
CHCCOPD1     0
ADDEPEV2     0
CHCKIDNY     0
DIABETE3     0
SEX          0
MARITAL      0
EMPLOY1      0
BLIND        0
DIFFWALK     0
ALCDAY5      0
_RFHYPE5     0
_MICHD       0
_DRDXAR1     0
_RACE       25
_AGEG5YR     0
HTM4         0
_EDUCAG      0
_SMOKER3     0
DRNKANY5     0
_RFBING5     0
_DRNKWEK     0
_FRTLT1      0
_VEGLT1      0
_PA150R2     0
_RFCHOL      0
WTKG3        0
dtype: int64

`_RACE` with NA cannot be assumed a value, therefore proceed to drop them (i.e. 25 records is 0.0025% of total data rows)

In [38]:
dataframe_v4.dropna(subset=['_RACE'], inplace = True)

In [39]:
dataframe_v4.isnull().sum()

CVDINFR4    0
CVDSTRK3    0
ASTHMA3     0
CHCSCNCR    0
CHCOCNCR    0
CHCCOPD1    0
ADDEPEV2    0
CHCKIDNY    0
DIABETE3    0
SEX         0
MARITAL     0
EMPLOY1     0
BLIND       0
DIFFWALK    0
ALCDAY5     0
_RFHYPE5    0
_MICHD      0
_DRDXAR1    0
_RACE       0
_AGEG5YR    0
HTM4        0
_EDUCAG     0
_SMOKER3    0
DRNKANY5    0
_RFBING5    0
_DRNKWEK    0
_FRTLT1     0
_VEGLT1     0
_PA150R2    0
_RFCHOL     0
WTKG3       0
dtype: int64

No more data with NA.  Proceed to check and transform data type, if required. 


In [40]:
dataframe_v4.describe().T

count         mean          std     min     25%           50%  \
CVDINFR4  933204.0     0.058666     0.234998     0.0     0.0  0.000000e+00   
CVDSTRK3  933204.0     0.041424     0.199269     0.0     0.0  0.000000e+00   
ASTHMA3   933204.0     0.135638     0.342404     0.0     0.0  0.000000e+00   
CHCSCNCR  933204.0     0.093551     0.291203     0.0     0.0  0.000000e+00   
CHCOCNCR  933204.0     0.097059     0.296038     0.0     0.0  0.000000e+00   
CHCCOPD1  933204.0     0.081627     0.273796     0.0     0.0  0.000000e+00   
ADDEPEV2  933204.0     0.192410     0.394194     0.0     0.0  0.000000e+00   
CHCKIDNY  933204.0     0.033823     0.180774     0.0     0.0  0.000000e+00   
DIABETE3  933204.0     0.136960     0.343806     0.0     0.0  0.000000e+00   
SEX       933204.0     0.415984     0.492891     0.0     0.0  0.000000e+00   
MARITAL   933204.0     0.521315     0.499546     0.0     0.0  1.000000e+00   
EMPLOY1   933204.0     0.490555     0.499911     0.0     0.0  0.000000e+00   
BLIND     933204.0     0.050578     0.219135     0.0     0.0  0.000000e+00   
DIFFWALK  933204.0     0.172671     0.377963     0.0     0.0  0.000000e+00   
ALCDAY5   933204.0     4.574287     7.854243     0.0     0.0  1.000000e+00   
_RFHYPE5  933204.0     0.404043     0.490706     0.0     0.0  0.000000e+00   
_MICHD    933204.0     0.088990     0.284730     0.0     0.0  0.000000e+00   
_DRDXAR1  933204.0     0.335643     0.472215     0.0     0.0  0.000000e+00   
_RACE     933204.0     2.003162     2.254127     1.0     1.0  1.000000e+00   
_AGEG5YR  933204.0     7.655081     3.431637     1.0     5.0  8.000000e+00   
HTM4      933204.0   169.369201    10.408297    91.0   163.0  1.693693e+02   
_EDUCAG   933204.0     1.269326     0.607647     0.0     1.0  1.000000e+00   
_SMOKER3  933204.0     0.681103     0.967774     0.0     0.0  0.000000e+00   
DRNKANY5  933204.0     0.485573     0.499792     0.0     0.0  0.000000e+00   
_RFBING5  933204.0     0.117268     0.321740     0.0     0.0  0.000000e+00   
_DRNKWEK  933204.0   330.503473   891.384366     0.0     0.0  5.397605e-79   
_FRTLT1   933204.0     0.572539     0.494710     0.0     0.0  1.000000e+00   
_VEGLT1   933204.0     0.706618     0.455312     0.0     0.0  1.000000e+00   
_PA150R2  933204.0     1.202589     0.826279     0.0     0.0  1.000000e+00   
_RFCHOL   933204.0     0.368106     0.482291     0.0     0.0  0.000000e+00   
WTKG3     933204.0  8048.080759  2028.685366  2268.0  6668.0  7.938000e+03   

                  75%      max  
CVDINFR4     0.000000      1.0  
CVDSTRK3     0.000000      1.0  
ASTHMA3      0.000000      1.0  
CHCSCNCR     0.000000      1.0  
CHCOCNCR     0.000000      1.0  
CHCCOPD1     0.000000      1.0  
ADDEPEV2     0.000000      1.0  
CHCKIDNY     0.000000      1.0  
DIABETE3     0.000000      1.0  
SEX          1.000000      1.0  
MARITAL      1.000000      1.0  
EMPLOY1      1.000000      1.0  
BLIND        0.000000      1.0  
DIFFWALK     0.000000      1.0  
ALCDAY5      4.574311     30.0  
_RFHYPE5     1.000000      1.0  
_MICHD       0.000000      1.0  
_DRDXAR1     1.000000      1.0  
_RACE        1.000000      9.0  
_AGEG5YR    10.000000     13.0  
HTM4       178.000000    241.0  
_EDUCAG      2.000000      2.0  
_SMOKER3     1.000000      3.0  
DRNKANY5     1.000000      1.0  
_RFBING5     0.000000      1.0  
_DRNKWEK   238.000000  53200.0  
_FRTLT1      1.000000      1.0  
_VEGLT1      1.000000      1.0  
_PA150R2     2.000000      2.0  
_RFCHOL      1.000000      1.0  
WTKG3     9072.000000  29030.0

In [41]:
dataframe_v4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 933204 entries, 0 to 933228
Data columns (total 31 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   CVDINFR4  933204 non-null  int64  
 1   CVDSTRK3  933204 non-null  int64  
 2   ASTHMA3   933204 non-null  int64  
 3   CHCSCNCR  933204 non-null  int64  
 4   CHCOCNCR  933204 non-null  int64  
 5   CHCCOPD1  933204 non-null  int64  
 6   ADDEPEV2  933204 non-null  int64  
 7   CHCKIDNY  933204 non-null  int64  
 8   DIABETE3  933204 non-null  int64  
 9   SEX       933204 non-null  int64  
 10  MARITAL   933204 non-null  int64  
 11  EMPLOY1   933204 non-null  int64  
 12  BLIND     933204 non-null  int64  
 13  DIFFWALK  933204 non-null  int64  
 14  ALCDAY5   933204 non-null  float64
 15  _RFHYPE5  933204 non-null  int64  
 16  _MICHD    933204 non-null  int64  
 17  _DRDXAR1  933204 non-null  int64  
 18  _RACE     933204 non-null  float64
 19  _AGEG5YR  933204 non-null  float64
 20  HTM4     

We see that there are some columns that we can define them as integers instead of float

In [42]:
dataframe_v4["ALCDAY5"] = dataframe_v4["ALCDAY5"].astype(int)
dataframe_v4["_RACE"] = dataframe_v4["_RACE"].astype(int)

In [43]:
dataframe_v4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 933204 entries, 0 to 933228
Data columns (total 31 columns):
 #   Column    Non-Null Count   Dtype  
---  ------    --------------   -----  
 0   CVDINFR4  933204 non-null  int64  
 1   CVDSTRK3  933204 non-null  int64  
 2   ASTHMA3   933204 non-null  int64  
 3   CHCSCNCR  933204 non-null  int64  
 4   CHCOCNCR  933204 non-null  int64  
 5   CHCCOPD1  933204 non-null  int64  
 6   ADDEPEV2  933204 non-null  int64  
 7   CHCKIDNY  933204 non-null  int64  
 8   DIABETE3  933204 non-null  int64  
 9   SEX       933204 non-null  int64  
 10  MARITAL   933204 non-null  int64  
 11  EMPLOY1   933204 non-null  int64  
 12  BLIND     933204 non-null  int64  
 13  DIFFWALK  933204 non-null  int64  
 14  ALCDAY5   933204 non-null  int32  
 15  _RFHYPE5  933204 non-null  int64  
 16  _MICHD    933204 non-null  int64  
 17  _DRDXAR1  933204 non-null  int64  
 18  _RACE     933204 non-null  int32  
 19  _AGEG5YR  933204 non-null  float64
 20  HTM4     

We will now rename our columns to readable format

In [44]:
new_names = {
    'CHCCOPD1': 'cpd_bronchitis',
    'ADDEPEV2': 'depression',
    '_DRDXAR1': 'arthritis',
    'CVDINFR4': 'heart_attack',
    'CVDSTRK3': 'stroke',
    'ASTHMA3': 'asthma',
    'DIABETE3': 'diabetes',
    'CHCKIDNY': 'kidney_disease',
    '_MICHD': 'heart_disease',
    'CHCSCNCR': 'skin_cancer',
    'CHCOCNCR': 'other_cancer',
    '_RFCHOL': 'high_cholesterol',
    '_RFHYPE5': 'high_bp',
    '_PA150R2': 'exercise_cat',
    'DRNKANY5': 'one_alc_per_day',
    'BLIND': 'blind',
    'MARITAL': 'martial',
    '_RFBING5': 'binge_drink',
    '_DRNKWEK': 'ave_drink_week',
    '_FRTLT1': 'fruit',
    '_VEGLT1': 'vegetable',
    'DIFFWALK': 'diff_walking',
    'ALCDAY5': 'occasion_drink_30days',
    'EMPLOY1': 'employment_status',
    '_SMOKER3': 'smoker_status',
    '_AGEG5YR': 'age',
    'SEX': 'sex',
    '_EDUCAG': 'education',
    '_RACE': 'race',
    'HTM4': 'height',
    'WTKG3': 'weight'
}

# Renaming the columns
dataframe_v4.rename(columns=new_names, inplace=True)

#### 5. Export Dataset for EDA
Export the cleaned dataframe into CSV - for Feature Engineering and EDA 

- `02_cleaned_data_forEDA.csv`


In [45]:
dataframe_v4.to_csv(f"../data/02_cleaned_data_forEDA.csv", index = False)